model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/idp/InfinispanIdentityProviderStorageProvider.java

```diff
@@ -140,6 +153,59 @@ public IdentityProviderModel getByAlias(String alias) {
         return createOrganizationAwareIdentityProviderModel(cached.getIdentityProvider());
     }
 
+    @Override
+    public Stream<IdentityProviderModel> getByOrganization(String orgId, Integer first, Integer max) {
+        RealmModel realm = getRealm();
+        String cacheKey = cacheKeyOrgId(realm, orgId);
+
+        // check if there is invalidation for this key or the organization was invalidated
+        if (isInvalid(cacheKey) || isInvalid(orgId)) {
+            return idpDelegate.getByOrganization(orgId, first, max).map(this::createOrganizationAwareIdentityProviderModel);
+        }
+
+        RealmCacheManager cache = realmCache.getCache();
+        IdentityProviderListQuery query = cache.get(cacheKey, IdentityProviderListQuery.class);
+        String searchKey = Optional.ofNullable(first).orElse(-1) + "." + Optional.ofNullable(max).orElse(-1);
+        Set<String> cached;
+
+        if (query == null) {
+            // not cached yet
+            Long loaded = cache.getCurrentRevision(cacheKey);
+            cached = idpDelegate.getByOrganization(orgId, first, max).map(IdentityProviderModel::getInternalId).collect(Collectors.toSet());
+            query = new IdentityProviderListQuery(loaded, cacheKey, realm, searchKey, cached);
+            cache.addRevisioned(query, startupRevision);
+        } else {
+            cached = query.getIDPs(searchKey);
+            if (cached == null) {
+                // there is a cache entry, but the current search is not yet cached
+                cache.invalidateObject(cacheKey);
+                Long loaded = cache.getCurrentRevision(cacheKey);
+                cached = idpDelegate.getByOrganization(orgId, first, max).map(IdentityProviderModel::getInternalId).collect(Collectors.toSet());
+                query = new IdentityProviderListQuery(loaded, cacheKey, realm, searchKey, cached, query);
+                cache.addRevisioned(query, cache.getCurrentCounter());
+            }
+        }
+
+        AtomicBoolean invalidate = new AtomicBoolean(false);
+        Stream<IdentityProviderModel> identityProviders = cached.stream()
+                .map((id) -> session.identityProviders().getById(id))
+                .takeWhile(idp -> {
+                    if (Objects.isNull(idp)) {
+                        invalidate.set(true);
+                        return false;
+                    }
+                    return true;
+                });
+
+        if (!invalidate.get()) {
```

Comment:
This won't work because the stream will not run until consumed after this method is called. This is the same issue we have in groups that I told you :)

Reply:
@pedroigor Changed to a for loop.